earth2mip/beta/perturbation/bv.py

```diff
@@ -0,0 +1,120 @@
+# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES.
+# SPDX-FileCopyrightText: All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+from typing import Callable
+
+import numpy as np
+import torch
+
+
+class BredVector:
+    """Bred Vector perturbation method, a classical technique for pertubations in
+    ensemble forecasting.
+
+    Parameters
+    ----------
+    model : Callable[[torch.Tensor], torch.Tensor]
+        Dynamical model, typically this is the prognostic AI model.
+        TODO: Update to prognostic looper
+    noise_amplitude : float, optional
+        Noise amplitude, by default 0.05
+    integration_steps : int, optional
+        Number of integration steps to use in forward call, by default 20
+    ensemble_perturb : bool, optional
+        Perturb the ensemble in an interacting fashion, by default False
+
+    Note
+    ----
+    For additional information:
+
+    - https://journals.ametsoc.org/view/journals/bams/74/12/1520-0477_1993_074_2317_efantg_2_0_co_2.xml
+    - https://en.wikipedia.org/wiki/Bred_vector
+    """
+
+    def __init__(
+        self,
+        model: Callable[
+            [torch.Tensor, OrderedDict[str, np.ndarray]],
+            tuple[torch.Tensor, OrderedDict[str, np.ndarray]],
+        ],
+        noise_amplitude: float = 0.05,
+        integration_steps: int = 20,
+        ensemble_perturb: bool = False,
+    ):
+        self.model = model
+        self.noise_amplitude = noise_amplitude
+        self.ensemble_perturb = ensemble_perturb
+        self.integration_steps = integration_steps
+        self.reddening = 2.0
+
+    @torch.inference_mode()
+    def __call__(
+        self,
+        x: torch.Tensor,
+        coords: OrderedDict[str, np.ndarray],
+    ) -> torch.Tensor:
+        """Apply perturbation method
+
+        Parameters
+        ----------
+        x : torch.Tensor
+            Input tensor intended to apply perturbation on
+        coords : OrderedDict[str, np.ndarray]
+            Ordered dict representing coordinate system that discribes the tensor
+
+        Returns
+        -------
+        torch.Tensor
+            Perturbation noise tensor
+        """
+        shape = x.shape
+        dx = self._generate_noise_correlated(shape, device=x.device)
```

Comment:
This can be made more generic and moved to instantiation.

```python
def __init__(self, 
       ...,
       seeding_perturbation_method: PerturbationMethod = Brown()
):

Reply:
Good call